src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp

```diff
@@ -20,86 +25,120 @@ class BankManager {
-    if (!tensor) {
-        OPENVINO_THROW("Uninitialized tensor in weights bank allocation!");
-    }
+void Bank::registerLT(const LazyTensor& tensor, const std::string& device) {
+    const std::string& device_for_alloc = m_alloc_device.empty() ? device : m_alloc_device;
+
+    std::lock_guard<std::mutex> guard(m_mutex);
 
-    if (device != "CPU" && device != "NPU") {
-        OPENVINO_THROW("Unsupported device in weights bank allocation: ", device);
+    auto& device_bank = m_device_bank[device_for_alloc];
+    if (device_bank.find(tensor) == device_bank.end()) {
+        device_bank[tensor] = ov::Tensor();
     }
+}
 
+void Bank::evaluate_and_allocate() {
     std::lock_guard<std::mutex> guard(m_mutex);
 
-    auto iter_cpu = m_bank.find(tensor.data());
-    if (iter_cpu == m_bank.end()) {
-        OPENVINO_THROW("Unknown tensor in weights bank allocation!");
+    for (auto&& bank : m_device_bank) {
+        const auto& device_for_alloc = bank.first;
+        auto& device_bank = bank.second;
+        std::vector<LazyTensor> vec;
+        for (const auto& el : device_bank) {
+            vec.push_back(el.first);
+        }
+        ov::parallel_for(vec.size(), [&](std::size_t idx) {
+            const auto& lt = vec[idx];
+            auto iter_device = device_bank.find(lt);
+            if (iter_device != device_bank.end() && iter_device->second) {
+                // Already allocated
+                return;
+            }
+
+            // Allocation and evaluation needed
+            unsafe_eval_and_alloc(lt, device_for_alloc);
+        });
     }
+}
 
-    // If target device is CPU - just reuse the default bank
-    if (device == "CPU") {
-        return iter_cpu->second;
+ov::Tensor Bank::unsafe_eval_and_alloc(const LazyTensor& tensor, const std::string& device_for_alloc) {
+    // Note: private method used inside other methods with already locked mutex
+    const auto& transformed_tensor = tensor.eval();
+    if (device_for_alloc == "CPU") {
+        m_device_bank[device_for_alloc][tensor] = transformed_tensor;
+        return transformed_tensor;
     }
 
-    // Non-CPU - check if the tensor is already there
-    auto& device_bank = m_device_bank[device];
-    auto iter_device = device_bank.find(tensor.data());
-    if (iter_device != device_bank.end()) {
-        // Already allocated on the device - reuse
-        return iter_device->second;
-    }
+    // FIXME: L0 allocation may crash when run in parallel
+    std::lock_guard<std::mutex> guard(m_alloc_mutex);
 
-    // Allocation needed
-    m_remote_ctx = m_core->get_default_context(device)._ptr;
-    auto remote_tensor = m_remote_ctx->create_host_tensor(tensor.get_element_type(), tensor.get_shape());
+    m_remote_ctx = m_core->get_default_context(device_for_alloc)._ptr;
+    auto remote_tensor =
+        m_remote_ctx->create_host_tensor(transformed_tensor.get_element_type(), transformed_tensor.get_shape());
```

Comment:
for the future: you can scope mutex right to call to `create_host_tensor` only. Now you have an expensive `copy_to` under the same mutex.

Reply:
Missed that, thanks! Added as a note for the next PR